yt_dlp/extractor/wasdtv.py

```diff
@@ -0,0 +1,169 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError,
+    int_or_none,
+    parse_iso8601,
+    urljoin,
+    try_get,
+)
+
+
+class WASDTVBaseIE(InfoExtractor):
+    _API_BASE = 'https://wasd.tv/api/'
+
+    def _fetch(self, path, video_id, description, query={}):
+        response = self._download_json(
+            urljoin(self._API_BASE, path),
+            video_id=video_id,
+            note=f'Downloading {description} metadata',
+            errnote=f'Unable to download {description} metadata',
+            query=query)
+        error = response.get('error')
+        if error:
+            raise ExtractorError(f'{self.IE_NAME} returned error: {error}', expected=True)
+        return response.get('result')
+
+    def _extract_thumbnails(self, thumbnails_dict):
+        if not thumbnails_dict:
+            return None
+        thumbnails = []
+        for index, thumbnail_size in enumerate(['small', 'medium', 'large']):
+            thumbnail_url = thumbnails_dict.get(thumbnail_size)
+            if not thumbnail_url:
+                continue
+            thumbnails.append({
+                'url': thumbnail_url,
+                'preference': index,
+            })
+        return thumbnails
+
+    def _real_extract(self, url):
+        container = self._get_container(url)
+        stream = try_get(container, lambda x: x['media_container_streams'][0])
+        media = try_get(stream, lambda x: x['stream_media'][0])
+        if not media:
+            raise ExtractorError('Can not extract media data.', expected=True)
+        media_meta = media.get('media_meta')
+        media_url, is_live = self._get_media_url(media_meta)
+        video_id = media.get('media_id') or container.get('media_container_id')
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(media_url, video_id, 'mp4')
+        self._sort_formats(formats)
+        return {
+            'id': str(video_id),
+            'title': container.get('media_container_name') or self._og_search_title(self._download_webpage(url, video_id)),
+            'description': container.get('media_container_description'),
+            'thumbnails': self._extract_thumbnails(media_meta.get('media_preview_images')),
+            'timestamp': parse_iso8601(container.get('created_at')),
+            'view_count': int_or_none(stream.get('stream_current_viewers' if is_live else 'stream_total_viewers')),
+            'is_live': is_live,
+            'formats': formats,
+            'subtitles': subtitles,
+        }
+
```

Comment:
```suggestion
    def _get_container(self, url):
        raise NotImplementedError('Must be defined in subclasses')

    def _get_media_url(self, media_meta):
        raise NotImplementedError('Must be defined in subclasses')

```
for self-documentation